django/queryset/README.md

```diff
@@ -0,0 +1,340 @@
+
+### Using `.exists()` to check if any data exists
+
+```python
+# Yes
+if Person.objects.filter(first_name='John').exists():
+    do_something()
+
+# No
+if Person.objects.filter(first_name='John'):
+    do_something()
+```
+
+Using `.exists()` consume less resources than not using it.
+
+### Using `.first()` instead of calling `.exists()`, followed by `.get()`
+
+```python
+# Yes
+matched_person = Person.objects.filter(employee_id='123').first()
+if matched_person:
+    do_something(matched_person)
+
+# No
+if Person.objects.filter(employee_id='123').exists():
+    matched_person = Person.objects.get(employee_id='123')
+    do_something(matched_person)
+```
+
+The reason is calling a single `.first()` will hit database only one. But when we call `.exists()`, it hits. Calling `.get()` will hits database again. Makes the second method to hit database twice which is a bad idea.
+
+### Avoid filtering across many relationship
+
+Filtering across _one-to-many_ or _many-to-many_ relationship can cause an unexpected result such as duplicated records.
+
+Suppose we have two departments
+
+```python
+hr = Department.objects.create(name='HRD')
+accounting = Department.objects.create(name='HRM')
+```
+
+John Doe is in both departments
+
+```python
+john_doe = Person.objects.create(...)
+hr.add(john_doe)
+accounting.add(john_doe)
+```
+
+If we try to get person who is in department that it's name contains word `HR`
+
+```python
+Person.objects.filter(departments__name__icontains='HR')
+# return: <QuerySet [<Person: John Doe>, <Person: John Doe>]>
+```
+
+We will get duplicated records because John Doe is in both departments. The solution for this case is to use `.distinct()`
+
+```python
+Person.objects.filter(departments__name__icontains='HR').distinct()
+# return: <QuerySet [<Person: John Doe>]>
+```
+
+### No need to use `.all()` before `.filter()`
+
+If you're using `.filter()` in your queryset, then you don't need to provide `.all()` anymore
+
+```python
+# Yes
+Person.objects.filter(first_name='John')
+
+# No
+Person.objects.all().filter(first_name='John')
+```
+
+### Avoid filtering against `JSONField`
+
+Filtering against `JSONField` can cause a huge performance degrades.
+
+If it is really needed to filter, then you might need to consider refactoring your model by using database relation (such as _one-to-many_ or _many-to-many_) to represent your data instead.
+
+### `Q() | Q()` is more preferred than union queryset
+
+```python
+# Yes
+queryset = Person.objects.filter(
+    Q(first_name='John') | Q(last_name='Doe')
+)
+
+# No
+queryset = Person.objects.filter(first_name='John')
+queryset.union(
+    Person.objects.filter(last_name='Doe')
+)
+```
+
+The first method will generate a query such as `SELECT ... FROM ... WHERE ... OR ...` which is faster than the second one which will generate `SELECT ... FROM ... WHERE ... UNION SELECT ... FROM ... WHERE ...`.
+
+## Manipulating Data
+
+### `.update()` is more preferred than looping save models
```

Comment:
Could you provide warning that `update()` is not triggering `post_save`?

Reply:
Added krubb.